sql/core/src/main/scala/org/apache/spark/sql/execution/BaseScriptTransformationExec.scala

```diff
@@ -47,7 +47,13 @@ trait BaseScriptTransformationExec extends UnaryExecNode {
   def ioschema: ScriptTransformationIOSchema
 
   protected lazy val inputExpressionsWithoutSerde: Seq[Expression] = {
-    input.map(Cast(_, StringType).withTimeZone(conf.sessionLocalTimeZone))
+    input.map { in =>
+      in.dataType match {
+        case _: ArrayType | _: MapType | _: StructType =>
+          new StructsToJson(in).withTimeZone(conf.sessionLocalTimeZone)
```

Comment:
How do users know we serialize complex type data using json format? This requires the external process know to parse json back to complext data, isn't? Is it well known to use json for similar feature in Hive?

Reply:
> How do users know we serialize complex type data using json format? This requires the external process know to parse json back to complext data, isn't? Is it well known to use json for similar feature in Hive?

In the first I implement it in same way as hive, then change to use json serde, you can see discussion in this part of comment https://github.com/apache/spark/pull/30957#discussion_r551727731

Reply:
Hmm, it's a new feature, so we can have a migration guide or release note make user know the new feature and difference.

Reply:
hm. Actually, this feature is not relevant to users using pre-built spark (w/ `-Phive` enabled)? IIUC there is no way to use this feature when enabling `-Phive`. I'm not sure about how important this feature is, but I think we need to document it somewhere if we will merge it.

Reply:
Yea， Hmmmm, your committer have some discuss result for this?

Reply:
> hm. Actually, this feature is not relevant to users using pre-built spark (w/ `-Phive` enabled)? IIUC there is no way to use this feature when enabling `-Phive`. I'm not sure about how important this feature is, but I think we need to document it somewhere if we will merge it.

That's why I implement this in hive's way first and support make json as a serde. If we implement this in json, then some origin hive SQL still can't run when directly upgrade to spark sql (w/ -Phive)。  How about add a configuration such as 
`spark.sql.sccriptTransform.useSparkFirst`